src/plugins/embeddable/public/react_embeddable_system/phase_tracker.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import {
+  PhaseEvent,
+  apiPublishesDataLoading,
+  apiPublishesRendered,
+} from '@kbn/presentation-publishing';
+import { BehaviorSubject, Subscription, combineLatest } from 'rxjs';
+
+export class PhaseTracker {
+  private firstLoadCompleteTime: number | undefined;
+  private embeddableStartTime = performance.now();
+  private subscriptions = new Subscription();
+  private phase$ = new BehaviorSubject<PhaseEvent | undefined>(undefined);
+
+  getPhase$() {
+    return this.phase$;
+  }
+
+  public trackPhaseEvents(uuid: string, api: unknown) {
+    const dataLoading$ = apiPublishesDataLoading(api)
+      ? api.dataLoading
+      : new BehaviorSubject(false);
+    const rendered$ = apiPublishesRendered(api) ? api.rendered$ : new BehaviorSubject(true);
```

Comment:
Sorry, quick question - should this be `hasRendered$`? No one actually seems to be implementing `PublishesRendered`, since the visualize embeddable calls it `hasRendered$` 🤔

Reply:
Thanks for pointing this out.

When designing this, I did not look at Visualize embeddable, so `PublishesRendered` API  and visualize embeddable `hasRendered$` are separate things. The visualize embeddable does not expose `hasRendered$` externally on its API.

This PR is designed to provide the API needed by lens to report rendering but can be used by other embeddables.

What do you recommend for Visualize embeddable? I could change `PublishesRendered` API to `hasRendered$` and update visualize embeddable to expose `hasRendered$` on its API. Thoughts?

Reply:
> I could change PublishesRendered API to hasRendered$ and update visualize embeddable to expose hasRendered$ on its API. Thoughts?

I like this! We could also continue to call it `rendered$`- either name is good for me. I tested the visualize embeddable and it seems to report `dataLoading$` as true followed by `rendered$`, which is exactly what we expect - so I think swapping it out would work? 🤔

Reply:
Resolved with https://github.com/elastic/kibana/pull/200640/commits/9bb2f8a3457efdbba2fe9aa038084a6e6c88a4bc